modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/api/ast/ArtifactAstUtils.java

```diff
@@ -107,43 +147,106 @@ public static ArtifactAst parseAndBuildAppExtensionModel(String[] configResource
               .build()));
     } else {
-      logModelNotGenerated("Mule ExtensionModelLoader not found", muleContext);
+      logModelNotGenerated("Mule ExtensionModelLoader not found", artifactId, ast.getArtifactType());
       return empty();
     }
   }
 
+  /**
+   * Handles a {@link ValidationResult} uniformly. Warnings are logged, errors raise exceptions.
+   *
+   * @param validationResult The {@link ValidationResult} obtained using an {@link ArtifactAstValidator}.
+   * @throws ConfigurationException If there was any result item with {@link ERROR} level.
+   */
+  public static void handleValidationResult(ValidationResult validationResult)
+      throws ConfigurationException {
+    final Collection<ValidationResultItem> items = validationResult.getItems();
+
+    // Logs warnings if any.
+    items.stream()
+        .filter(v -> v.getValidation().getLevel().equals(WARN))
+        .forEach(v -> LOGGER.warn(validationResultItemToString(v)));
+
+    // If there are any errors, throws.
+    final boolean hasErrors = items.stream().anyMatch(v -> v.getValidation().getLevel().equals(ERROR));
+    if (hasErrors) {
+      throw new ConfigurationException(createStaticMessage(validationResult.getItems()
+          .stream()
+          .map(ArtifactAstUtils::validationResultItemToString)
+          .collect(joining(lineSeparator()))));
+    }
+  }
+
+  private static String validationResultItemToString(ValidationResultItem v) {
+    return v.getComponents().stream()
+        .map(component -> component.getMetadata().getFileName().orElse("unknown") + ":"
+            + component.getMetadata().getStartLine().orElse(-1))
+        .collect(joining("; ", "[", "]")) + ": " + v.getMessage();
+  }
+
   private static ArtifactAst doParseArtifactIntoAst(String[] configResources,
                                                     AstXmlParserSupplier parserSupplier,
                                                     Set<ExtensionModel> extensions,
-                                                    boolean disableValidations) {
-    try {
-      return parserSupplier.getParser(extensions, disableValidations).parse(loadConfigResources(configResources));
-    } catch (Exception e) {
-      throw new MuleRuntimeException(e);
+                                                    boolean disableValidations)
+      throws ConfigurationException {
+    return parserSupplier.getParser(extensions, disableValidations).parse(loadConfigResources(configResources));
+  }
+
+  private static boolean definesOwnExtensionModel(ArtifactAst artifactAst) {
+    ArtifactType artifactType = artifactAst.getArtifactType();
+    switch (artifactType) {
+      case APPLICATION:
+        return artifactAst.topLevelComponentsStream()
+            .anyMatch(component -> REUSABLE_COMPONENT_TYPES.contains(component.getComponentType()));
+      case MULE_EXTENSION:
+        // TODO W-11796932: Try to unify this one with the case above by changing the Mule Extensions DSL
```

Comment:
I don't see the need for this unification as they're conceptually different things that can (and most likely will) evolve separetly.

Reply:
I agree, but I'm suggesting to do it not because we **need to** but because it could make our lives easier if we can.
If there comes the time in which we can't do that anymore, we can add complexity then, so long as the API supports it.
I've also created this other task: ([W-11796759](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE0000170sjtYAA/view)) to move this logic into the `mule-extensions-support` module. Also added the TODO [here](https://github.com/mulesoft/mule/pull/11776/files#diff-bdcfd9b812bf50e66a6a7978ac2c90666ce05ab08ae2a57328b7cbfeb168b78cR60).

Reply:
My point is that I don't see how would that make your life any easier, and if it does, it will be short lived. Thus, the TODO only enlarges an already large backlog with a task of little to none value.

As for the second TODO, this shouldn't move to `mule-module-extension-support` but to `mule-module-extensions-mule-support`

Reply:
Yes, my bad, I meant to say `mule-module-extensions-mule-support`.

I just removed the `TODO` with the rationale that the logic is fine as it is. The task [W-11796932](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE0000170jmbYAA/view) may or may not be done, but if it is done, it shouldn't be because of this `TODO`. In case we do it, we may want to come back and simplify this part, but not necessarily.

We can also revisit it when we do [W-11796759](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE0000170sjtYAA/view).